controllers/complianceeventsapi/auth.go

```diff
@@ -0,0 +1,186 @@
+			Insecure:   cfg.TLSClientConfig.Insecure,
+		},
+		BearerToken: token,
+	}
+
+	return kubernetes.NewForConfig(userConfig)
+}
+
+// canRecordComplianceEvent will perform certificate or token authentication and perform a subject access review to
+// ensure the input user has patch access to patch the policy status in the managed cluster namespace. An error is
+// returned if the authorization could not be determined. Note that authenticatedClient and authenticator can be nil
+// if certificate authentication isn't used. If both certificate and token authentication is present, certificate takes
+// precedence.
+func canRecordComplianceEvent(
+	cfg *rest.Config,
+	authenticatedClient *kubernetes.Clientset,
+	authenticator *apiserverx509.Authenticator,
+	clusterName string,
+	req *http.Request,
+) (bool, error) {
+	postRules := authzv1.ResourceAttributes{
+		Group:       "policy.open-cluster-management.io",
+		Version:     "v1",
+		Resource:    "policies",
+		Verb:        "patch",
+		Namespace:   clusterName,
+		Subresource: "status",
+	}
+
+	// req.TLS.PeerCertificates will be empty if certificate authentication is not enabled (e.g. endpoint is not HTTPS)
+	if req.TLS != nil && len(req.TLS.PeerCertificates) > 0 {
+		resp, ok, err := authenticator.AuthenticateRequest(req)
+		if err != nil {
+			if errors.As(err, &x509.UnknownAuthorityError{}) || errors.As(err, &x509.CertificateInvalidError{}) {
+				return false, ErrNotAuthorized
+			}
+
+			return false, err
+		}
+
+		if !ok {
+			return false, nil
+		}
+
+		review, err := authenticatedClient.AuthorizationV1().SubjectAccessReviews().Create(
+			req.Context(),
+			&authzv1.SubjectAccessReview{
+				Spec: authzv1.SubjectAccessReviewSpec{
+					ResourceAttributes: &postRules,
+					User:               resp.User.GetName(),
+					Groups:             resp.User.GetGroups(),
+					UID:                resp.User.GetUID(),
+				},
+			},
+			metav1.CreateOptions{},
+		)
+		if err != nil {
+			return false, err
+		}
+
+		if !review.Status.Allowed {
+			log.V(0).Info(
+				"The user is not authorized to record a compliance event",
+				"cluster", clusterName,
+				"user", resp.User.GetName(),
+			)
+		}
+
+		return review.Status.Allowed, nil
+	}
+
+	token := parseToken(req)
+	if token == "" {
+		return false, ErrNotAuthorized
+	}
+
+	userClient, err := getClientFromToken(cfg, token)
+	if err != nil {
+		return false, err
+	}
+
+	result, err := userClient.AuthorizationV1().SelfSubjectAccessReviews().Create(
+		req.Context(),
+		&authzv1.SelfSubjectAccessReview{
+			Spec: authzv1.SelfSubjectAccessReviewSpec{
+				ResourceAttributes: &postRules,
+			},
+		},
+		metav1.CreateOptions{},
+	)
+	allowed := false
+
+	if err != nil {
+		if !k8serrors.IsUnauthorized(err) {
+			return false, ErrNotAuthorized
+		}
+	} else {
+		allowed = result.Status.Allowed
+	}
+
+	if !allowed {
+		var username string
+
+		userInfo, err := getTokenUserInfo(req.Context(), authenticatedClient, token)
```

Comment:
This is just for log.. do we really need?

Reply:
I changed it to try parsing the token locally as a best effort. It's important to log unauthorized access attempts for auditing.